### PR TITLE
Add Flash Attention 2 support to Bark

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Following a recent series of PRs and issues to improve Bark, this PR aims to add FA2 support to Bark.  Bark self-attention class supports both causal and non-causal attention but otherwise changes are minimal.

I've also taken the opportunity to switch to  `_prepare_4d_attention_mask` instead of manually creating the 4d attention mask.

Benchmarks are currently running at the moment to measure speed/memory gains!

cc @sanchit-gandhi and @amyeroberts

### CODE REVIEW

File: src/transformers/models/bark/modeling_bark.py [hunk 1]
```diff
@@ -1669,3 +1856,28 @@ def generate(
             return audio, output_lengths
 
         return audio
+
+    @classmethod
+    def _check_and_enable_flash_attn_2(
+        cls, config, torch_dtype: Optional[torch.dtype] = None, device_map: Optional[Union[str, Dict[str, int]]] = None
+    ):
+        """
+        If you don't know about Flash Attention, check out the official repository of flash attention:
```
Comment (sanchit-gandhi) [inline src/transformers/models/bark/modeling_bark.py:1865]: Could be worth explaining quickly why we override this method in the docstring!

File: tests/models/bark/test_modeling_bark.py [hunk 2]
```diff
@@ -872,6 +875,160 @@ def test_resize_embeddings_untied(self):
             # Check that the model can still do a forward pass successfully (every parameter should be resized)
             model(**self._prepare_for_class(inputs_dict, model_class))
 
+    @require_flash_attn
+    @require_torch_gpu
+    @mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_inference(self):
+        import torch
```
Comment (sanchit-gandhi) [inline tests/models/bark/test_modeling_bark.py:883]: (nit) no need for this import as we already gate it at the start of the tester file
```suggestion
```

File: tests/models/bark/test_modeling_bark.py [hunk 3]
```diff
@@ -872,6 +875,160 @@ def test_resize_embeddings_untied(self):
...
+                )
+                model.to(torch_device)
+
+                dummy_input = inputs_dict["input_ids"][:1]
+                if dummy_input.dtype in [torch.float32, torch.float16]:
+                    dummy_input = dummy_input.to(torch.bfloat16)
+
+                dummy_attention_mask = inputs_dict.get("attention_mask", None)
+
+                if dummy_attention_mask is not None:
```
Comment (sanchit-gandhi) [inline tests/models/bark/test_modeling_bark.py:910]: What's the motivation behind overriding the attention mask here?

File: tests/models/bark/test_modeling_bark.py [hunk 4]
```diff
@@ -872,6 +875,160 @@ def test_resize_embeddings_untied(self):
...
+                    dummy_attention_mask = dummy_attention_mask[:1]
+                    dummy_attention_mask[:, 1:] = 1
+                    dummy_attention_mask[:, :1] = 0
+
+                outputs = model(inputs_dict["codebook_idx"], dummy_input, output_hidden_states=True)
+                outputs_fa = model_fa(inputs_dict["codebook_idx"], dummy_input, output_hidden_states=True)
+
+                logits = (
+                    outputs.hidden_states[-1]
+                    if not model.config.is_encoder_decoder
```
Comment (sanchit-gandhi) [inline tests/models/bark/test_modeling_bark.py:920]: We know bark is not encoder-decoder -> could we simplify the tests to reflect this?

File: tests/models/bark/test_modeling_bark.py [hunk 5]
```diff
@@ -872,6 +875,160 @@ def test_resize_embeddings_untied(self):
...
+                    else outputs.decoder_hidden_states[-1]
+                )
+                logits_fa = (
+                    outputs_fa.hidden_states[-1]
+                    if not model.config.is_encoder_decoder
+                    else outputs_fa.decoder_hidden_states[-1]
+                )
+
+                assert torch.allclose(logits_fa, logits, atol=4e-2, rtol=4e-2)
```
Comment (sanchit-gandhi) [inline tests/models/bark/test_modeling_bark.py:929]: Pretty high tolerance! We've compared the audio outputs qualitatively with / without flash attention and they match?

File: tests/models/bark/test_modeling_bark.py [hunk 6]
```diff
@@ -872,6 +875,160 @@ def test_resize_embeddings_untied(self):
...
+                # check with inference + dropout
+                model.train()
+                _ = model_fa(inputs_dict["codebook_idx"], dummy_input, **other_inputs)
+
+    @require_flash_attn
+    @require_torch_gpu
+    @mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_inference_padding_right(self):
+        import torch
```
Comment (sanchit-gandhi) [inline tests/models/bark/test_modeling_bark.py:962]:
```suggestion
```

**Review by sanchit-gandhi — APPROVED** [review]
Very clean - thanks @ylacombe for adding this! Keen to see what kind of performance gain we get from this

File: tests/models/bark/test_modeling_bark.py (hunk 5 above)
Comment (ylacombe) [reply tests/models/bark/test_modeling_bark.py:929]: Thanks for the quick review, I've actually copied out and modified a test that is in the general suite, so I haven't change anything -> tolerance and attention mask overriding are the same than the original test!

File: tests/models/bark/test_modeling_bark.py (hunk 3 above)
Comment (ylacombe) [reply tests/models/bark/test_modeling_bark.py:909]: Making sure that at least one of the input ids is masked !

**Comment by ylacombe** [comment]
Thanks for the quick review, I've addressed your comments :hugs:

**Review by amyeroberts — APPROVED** [review]
Very nice - thanks for adding!

**Comment by ylacombe** [comment]
Merging ! thanks for the quick reviews!

**Review by ArthurZucker — COMMENTED** [review]
Thanks for adding this! I usually also request to add a section in the readme, and update the flash attention list of models that are supported [here](https://github.com/huggingface/transformers/pull/26479/files#diff-2a85c68b1238c345068cca421d21974fe8e478d94eb3931cd18fa364391122e6) and the readme like [this change](https://github.com/huggingface/transformers/pull/26479/files#diff-9a9d8a0eb031f73b3a7a2141b1855d21c40a55b94d66a6c6657d679d502ed620).

File: src/transformers/models/bark/modeling_bark.py [hunk 7]
```diff
@@ -187,6 +207,177 @@ def forward(
...
+            key = torch.cat((past_key, key), dim=1)
+            value = torch.cat((past_value, value), dim=1)
+
+        if use_cache is True:
+            #  (batch, head, seq_length, head_features)
+            present = (key.transpose(1, 2), value.transpose(1, 2))
+        else:
+            present = None
+
+        attn_output = self._flash_attention_forward(query, key, value, attention_mask, query_len, dropout=self.dropout)
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/models/bark/modeling_bark.py:268]: It's actually a float here: https://github.com/ylacombe/transformers/blob/3258ff93304078b9e27d752e6c19d3813f664855/src/transformers/models/bark/modeling_bark.py#L93-L94 !
